Teach Bogofilter spam mails that were detected by Rspamd only

matthias-g opened the pull request:
Mails that are classified by Rspamd as spam but not by Bogofilter are treated as spam. This small change teaches Bogofilter that such mails are spam to improve Bogofilter's accuracy.

The logic changes as follows: If mails are classified as spam by Rspamd only, they do not get the "X-Spam-Folder: YES" header. After they have been read and left in the spam folder, they are learned by Bogofilter as spam. If mails are not marked as read or moved to another folder directly after reading, nothing happens.
Thus, to teach Bogofilter, one has to mark correctly classified spam mails as read.

matthias-g committed:
Teach Bogofilter spam mails that were detected by Rspamd only

bhopmann commented:
@matthias-g Many thanks for your PR.

Two thoughts about it (feel free to correct me if I might be wrong thought it's been a while I've written the script):
- This probably interferes with the option `SILENT_NEW_SPAM` - when set to `1` emails are always marked as read and teaching spam via bogofilter (via mark as read) isn't controlable any more:
https://github.com/bhopmann/uberspace7-mailfilter/blob/09f7d338cc36eb9eb7fea00df8bea2c83495bf61/.mailfilters/handlespam#L119
- When Rspamd classifies emails as false negatives, bogofilter does the same thing

matthias-g committed:
Do not use mails detected by Rspamd to train Bogofilter if SILENT_NEW_SPAM is set

The training takes mails marked as seen as correctly classified spam.
This interferes with SILENT_NEW_SPAM marking all spam as seen.

matthias-g commented:
Thanks for the reply!

> This probably interferes with the option SILENT_NEW_SPAM - when set to 1 emails are always marked as read and teaching spam via bogofilter (via mark as read) isn't controlable any more

Yes, I've missed that. SILENT_NEW_SPAM seems to be inherently incompatible with this PR's mechanism. I modified the PR so that the additional training for bogofilter is not used in this case, i.e. if SILENT_NEW_SPAM is set, the PR does not change the current behaviour.

> When Rspamd classifies emails as false negatives, bogofilter does the same thing

Can you please elaborate on that? From my understanding, if Rspamd classifies an email as false negative, bogofilter might still flag it as true positive and the email will be moved to the spam folder.

bhopmann commented:
Sorry for my late reply! 

> > When Rspamd classifies emails as false negatives, bogofilter does the same thing

> Can you please elaborate on that? From my understanding, if Rspamd classifies an email as false negative, bogofilter might still flag it as true positive and the email will be moved to the spam folder.

The main idea when writing the script was to let Rspamd do the basic filtering job and to have bogofilter for custom spam or ham filtering. Therefore bogofilter was added to correct Rspamd, but only in case Rspamd sorted mails as false negative or false positive. When Rspamd sorts mails as spam (and one finds that the result is indeed spam), bogofilter isn't addressed because the result is wanted. Only, if spam isn't found by Rspamd (false positive), bogofilter can be teached by moving the mail in spam folder. And vice versa, bogofilter is teached, when mails are incorrectly classified as spam (false negative) - in this case, mails are correctly classified by bogofilter by moving them into the inbox. This all was done with simplicity in mind (without extra moves or any custom folders). 

However, it seems to be a good idea, to emancipate bogofilter from Rspamd: I'm currently running my setup with your additions for testing purposes. It works very nicely so far!

By the way, setting a `SPAM_SUBJECT_TAG` in `bogofilter.cf` allows identifying all mails, bogofilter hasn't classified (so they can be marked as read and consequently classified as spam manually). Otherwise, one has to look in the email header. Or mark every incoming junk as read.

```
#### SPAM_SUBJECT_TAG
#
#	tag added to "Subject: " line for identifying spam or unsure
#	default is to add nothing.
#
spam_subject_tag=***SPAM***
unsure_subject_tag=???UNSURE???
```
